nx/lib/nx.ex

```diff
@@ -14312,7 +14312,27 @@ defmodule Nx do
 
     shape = Nx.Shape.take_along_axis(tensor.shape, indices.shape, axis)
 
-    result = impl!(tensor).take_along_axis(%{tensor | shape: shape}, tensor, indices, axis)
+    out = %{tensor | shape: shape}
+
+    result =
+      Nx.Shared.optional(:take_along_axis, [tensor, indices, axis], out, fn tensor,
+                                                                            indices,
+                                                                            axis ->
+        axes_range = axes(indices)
+        new_axis_shape = Tuple.append(shape(indices), 1)
+
+        full_indices =
+          axes_range
+          |> Enum.map(fn
+            ^axis -> reshape(indices, new_axis_shape)
+            axis -> iota(new_axis_shape, axis: axis)
+          end)
+          |> concatenate(axis: rank(indices))
+
+        out = gather(tensor, full_indices)
+
+        %{out | shape: shape}
```

Comment:
Interesting, why do you need to change the shape? We should probably avoid doing this!

Reply:
I just copied it from the deleted impl call without thinking much about it.